### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by Andyck**
PR contains:
- Refactoring of delete methods in CloudOfThingsRestClient.java
- Remove workaround with 'x'-Parameter using appropriate deleteBy-method

### CODE REVIEW

File: src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java [hunk 1]
```diff
@@ -607,47 +607,14 @@ public String doPutRequestWithIdResponse(String data, String path, String conten
...
-                throw new CotSdkException(response.code(), "Error in delete with ID '" + id + "' (see https://http.cat/" + response.code() + ")");
-            }
-        } catch( CotSdkException e) {
-            // We need to rethrow this in order to not loose the status code.
-            throw e;
-        } catch (Exception e) {
-            throw new CotSdkException("Error in request", e);
-        } finally {
-            closeResponseBodyIfResponseAndBodyNotNull(response);
+        if(id == null || id.isEmpty()) {
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java]: :+1: It's good to fail early.

File: src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java [hunk 2]
```diff
@@ -662,8 +629,11 @@ public void delete(String url) {
             if (!response.isSuccessful()) {
                 throw new CotSdkException(response.code(), "Error in delete with URL '" + url + "' (see https://http.cat/" + response.code() + ")");
             }
+        } catch( CotSdkException e) {
+            // We need to rethrow this in order to not loose the status code.
+            throw e;
         } catch (Exception e) {
-            throw new CotSdkException("Error in request", e);
+            throw new CotSdkException("Error in request: " + e.getMessage(), e);
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java]: :+1:

File: src/main/java/com/telekom/m2m/cot/restsdk/alarm/AlarmApi.java [hunk 3]
```diff
@@ -113,15 +113,15 @@ public AlarmCollection getAlarms(@Nullable Filter.FilterBuilder filters, int res
...
     public void deleteAlarms(@Nullable final Filter.FilterBuilder filters) {
         if(filters != null) {
             filters.validateSupportedFilters(acceptedFilters);
         }
         final String filterParams = Optional.ofNullable(filters)
-            .map(filterBuilder -> filterBuilder.buildFilter() + "&")
+            .map(filterBuilder -> filterBuilder.buildFilter())
             .orElse("");
-        // The x query parameter is a workaround. Without, it seems as if there are cases where deletion does not work.
-        cloudOfThingsRestClient.delete("", RELATIVE_API_URL + "?" + filterParams + "x=");
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/alarm/AlarmApi.java]: Good to see that this hack has been removed :+1:

File: src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java [hunk 4]
```diff
@@ -607,47 +607,14 @@ public String doPutRequestWithIdResponse(String data, String path, String conten
...
-            }
-        } catch (CotSdkException e) {
-            // We need to rethrow this in order to not loose the status code.
-            throw e;
-        } catch (Exception e) {
-            throw new CotSdkException("Error in request: " + e.getMessage(), e);
-        } finally {
-            closeResponseBodyIfResponseAndBodyNotNull(response);
-        }
+        delete(host + "/" + trimSlashes(api) + "?" + filter);
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java]: What if the filter is empty? Does it delete all objects and did we test that?

File: src/main/java/com/telekom/m2m/cot/restsdk/event/EventApi.java [hunk 5]
```diff
@@ -108,11 +110,15 @@ public EventCollection getEvents(Filter.FilterBuilder filters) {
...
-    public void deleteEvents(Filter.FilterBuilder filters) {
+    public void deleteEvents(@Nullable final Filter.FilterBuilder filters) {
         if(filters != null) {
             filters.validateSupportedFilters(acceptedFilters);
         }
-        cloudOfThingsRestClient.delete("", RELATIVE_API_URL + "?" + filters.buildFilter() + "&x=");
+        final String filterParams = Optional.ofNullable(filters)
+                .map(filterBuilder -> filterBuilder.buildFilter())
+                .orElse("");
+        cloudOfThingsRestClient.deleteBy(filterParams, RELATIVE_API_URL);
```
Comment (Matthimatiker) [inline src/main/java/com/telekom/m2m/cot/restsdk/event/EventApi.java]: :+1:

File: src/test/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClientTest.java [hunk 6]
```diff
@@ -91,7 +91,7 @@ public void testDeleteWithException() throws Exception {
 
         CloudOfThingsRestClient cloudOfThingsRestClient = new CloudOfThingsRestClient(clientMock, TEST_HOST, TEST_USERNAME, TEST_PASSWORD);
 
-        cloudOfThingsRestClient.delete("", "");
+        cloudOfThingsRestClient.delete("id", "api");
```
Comment (Matthimatiker) [inline src/test/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClientTest.java]: :+1:

File: src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java (hunk 4 above)
Comment (Andyck) [reply src/main/java/com/telekom/m2m/cot/restsdk/CloudOfThingsRestClient.java]: I tested it: when filter is empty all objects will be deleted as mentioned in javadoc.

**Pull request merged**